src/ccl_tracers.c

```diff
@@ -167,38 +204,239 @@ static double lensing_kernel_integrand(double z, void *pars) {
+            *status = CCL_ERROR_INTEG;
+            ccl_cosmology_set_status_message(
+              cosmo,
+              "ccl_tracers.c: integrate_lensing_kernel_gsl(): error in computing lensing kernel. "
+              "Integration error at chi %g larger than tolerance of %g: %g\n",
+              chi_arr[i],  wL_max*rel_tol, wL_err_arr[i]);
+            break;
+          }
+        }
+      }
+    }
+  }
+  free(wL_err_arr);
+}
+
+// Computes the lensing kernel integral using spline integration:
+// 3 * H0^2 * Omega_M / 2 / a *
+// Integral[ p(z) * (1-5s(z)/2) * chi_end * (chi(z)-chi_end)/chi(z) ,
+//          {z',z_end,z_max} ]
+static void integrate_lensing_kernel_spline(ccl_cosmology *cosmo,
+                                              int nz, double* z_arr, double* nz_arr, double nz_norm,
+                                              ccl_f1d_t* sz_f,
+                                              int nchi, double* chi_arr, double* wL_arr,
+                                              int* status) {
+  double* chi_of_z_array = malloc(nz*sizeof(double));
+  double* sz_array = malloc(nz*sizeof(double));
+  if(chi_of_z_array == NULL || sz_array == NULL) {
+    *status = CCL_ERROR_MEMORY;
+    ccl_cosmology_set_status_message(
+      cosmo,
+      "ccl_tracers.c: lensing_kernel_integrate_spline(): error allocating memory\n");
+  }
+
+  if(*status == 0) {
+    // Fill chi and sz arrays
+    for(int i=0; i<nz; i++) {
+      double a = 1./(1+z_arr[i]);
+      chi_of_z_array[i] = ccl_comoving_radial_distance(cosmo, a, status);
+      if(sz_f != NULL) {
+        sz_array[i] = ccl_f1d_t_eval(sz_f, z_arr[i]);
+      } else {
+        sz_array[i] = 1.0;
+      }
+    }
+  }
+
+  if(*status == 0) {
+    #pragma omp parallel default(none) \
+                        shared(cosmo, nz, z_arr, nz_arr, nz_norm, sz_f, \
+                               chi_of_z_array, sz_array, \
+                               nchi, chi_arr, wL_arr, status, gsl_interp_akima)
+    {
+      int local_status = *status;
+      double lens_prefac = get_lensing_prefactor(cosmo, &local_status);
+      double result = 0.0;
+      double chi, chi_end, a, z_end;
+
+      double *integrand_array = malloc(nz*sizeof(double));
+      if(integrand_array == NULL) {
+        local_status = CCL_ERROR_MEMORY;
+        ccl_cosmology_set_status_message(
+          cosmo,
+          "ccl_tracers.c: lensing_kernel_integrate_spline(): error allocating memory\n");
+      }
+      if(local_status == 0) {
+        #pragma omp for
+        for(int ichi=0; ichi<nchi; ichi++) {
+          chi_end = chi_arr[ichi];
+          a = ccl_scale_factor_of_chi(cosmo, chi_end, &local_status);
+          z_end = 1./a-1;
+
+          // We don't need to start at 0 but finding the index corresponding to chi_end would make things more complicated
+          for(int i=0; i<nz; i++) {
+            if(chi_of_z_array[i] < chi_end) {
+              integrand_array[i] = 0.0;
```

Comment:
OK, if I understand this correctly, this will work well assuming that the N(z) is more finely sampled than the lensing kernel, correct? Is there a real-world situation where this may not be the case?

Reply:
That's a good point. If the n(z) is sampled very sparsely, then this could become inaccurate. In that case the quadrature methods would work better. One option would be to choose quadrature when `n_chi` > `nz.size`, while allowing the user to force either spline integration or quadrature.

Reply:
How about just complementing the spline integral with a single trapezoidal integral for the piece between `chi_end` and `chi_of_z_array[i_first]`, where `chi_of_z_array[i_first]` is the first value of  `chi_of_z_array` satisfying `chi_of_z_array[i_first] >= chi_end`?

Reply:
It'd require assuming something about what `Nz(chi_end)` is, but this could be calculated from the spline interpolator.

Reply:
The integrand at chi_end is 0, so that was simple to account for. There's a subtlety when z[0] >> 0 and nz[0] > 0 but that was luckily caught by a test (and I added another one for that specific edge case).